mhs-reference-implementation/mhs/parser/ebxml_message_parser.py

```diff
@@ -1,38 +1,137 @@
+
+EBXML_NAMESPACE = "eb"
+SOAP_NAMESPACE = "SOAP"
+NAMESPACES = {SOAP_NAMESPACE: "http://schemas.xmlsoap.org/soap/envelope/",
+              EBXML_NAMESPACE: "http://www.oasis-open.org/committees/ebxml-msg/schema/msg-header-2_0.xsd"}
+
+CONTENT_TYPE_HEADER_NAME = "Content-Type"
+
+
+class EbXmlParsingError(Exception):
+    """Raised when an error was encountered during parsing of an ebXML message."""
+    pass
 
 
 class EbXmlMessageParser:
     """A component that extracts information from ebXML messages."""
 
-    def parse_message(self, message):
+    def parse_message(self, headers, message):
         """Parse the provided ebXML message and extract a dictionary of values from it.
 
+        :param headers A dictionary of headers received with the message.
         :param message: The message to be parsed.
         :return: A dictionary of values extracted from the message.
         """
-        # TODO: Implement schematron/XSLT parsing here.
-        pass
+        xml_tree = ET.fromstring(message)
+        extracted_values = {}
 
+        self._extract_ebxml_text_value(xml_tree, "PartyId", extracted_values, FROM_PARTY_ID, parent="From")
+        self._extract_ebxml_text_value(xml_tree, "PartyId", extracted_values, TO_PARTY_ID, parent="To")
+        self._extract_ebxml_text_value(xml_tree, "CPAId", extracted_values, CPA_ID)
+        self._extract_ebxml_text_value(xml_tree, "ConversationId", extracted_values, CONVERSATION_ID)
+        self._extract_ebxml_text_value(xml_tree, "Service", extracted_values, SERVICE)
+        self._extract_ebxml_text_value(xml_tree, "Action", extracted_values, ACTION)
+        self._extract_ebxml_text_value(xml_tree, "MessageId", extracted_values, MESSAGE_ID, parent="MessageData")
+        self._extract_ebxml_text_value(xml_tree, "Timestamp", extracted_values, TIMESTAMP, parent="MessageData")
+        self._extract_ebxml_text_value(xml_tree, "RefToMessageId", extracted_values, REF_TO_MESSAGE_ID,
+                                       parent="MessageData")
+        self._extract_ebxml_value_is_present(xml_tree, "DuplicateElimination", extracted_values, DUPLICATE_ELIMINATION)
+        self._extract_ebxml_value_is_present(xml_tree, "SyncReply", extracted_values, SYNC_REPLY)
+        self._extract_ebxml_value_is_present(xml_tree, "AckRequested", extracted_values, ACK_REQUESTED)
+        self._extract_attribute(xml_tree, "AckRequested", SOAP_NAMESPACE, "actor", extracted_values, ACK_SOAP_ACTOR)
 
-class EbXmlRequestMessageParser(EbXmlMessageParser):
-    """A component that extracts information from ebXML acknowledgement messages."""
+        return extracted_values
+
+    def _path_to_ebxml_element(self, name, parent=None):
+        path = ".//"
+
+        if parent is not None:
+            path += EBXML_NAMESPACE + ":" + parent + "/"
+
+        path += EBXML_NAMESPACE + ":" + name
+
+        return path
+
+    def _extract_ebxml_text_value(self, xml_tree, element_name, values_dict, key, parent=None):
```

Comment:
It looks like this function is doing 2 things. It is extracting the value from the xml tree and also update the values_dict with that value. Do you think there is any value in separating this out? or perhaps renaming this function.

Reply:
I think you're right, I could extract the dictionary update to a separate function to make things clearer. Will update.

Reply:
I've refactored this. Hopefully it makes things a little clearer.